### PR TITLE
[engine] importValue with upgrade

### PR DESCRIPTION
**Pull request by nickchapman-da**
Advances #17082

Implement the compatibility relation (i.e code to drop/make-None) within `importValue`:

- Pass: `optTargetTemplateId: Option[TypeConName]` to `SBFetchAny`
- And pass this target type when supplied to `importValue`, which now does up/down-grade
- `SBCastAnyContract` now reverts to previous simpler behaviour

Also:
- avoid passing `fields` throughout `Compiler.scala` (no longer needed)
- and remove the MVP `Casting` abstraction

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -191,7 +191,7 @@ private[speedy] sealed abstract class SBuiltin(val arity: Int) {
       i: Int,
   ): (TypeConName, SRecord) =
     args.get(i) match {
-      case SAnyContract(tyCon, value) => (tyCon, value)
+      case SAny(_, record: SRecord) => (record.id, record)
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:194]:
```suggestion
      case SAny(TyCon(tyCon), record: SRecord) => (tyCon, record)
```
Please do not rely on the type in the record.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 2]
```diff
@@ -123,15 +124,16 @@ private[lf] object Speedy {
 
   final case class CachedContract(
       version: TxVersion,
-      templateId: Ref.TypeConName,
-      value: SValue,
+      templateId: Ref.TypeConName, // TODO: dont pass; recover from (record)value.
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:127]: please do not rely on Record type.

**Comment by remyhaemmerle-da** [comment]
I am curious to know how you will handle the caching of contract when a same contract need to be interpreted with different versions in the same transactions.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:194]: ok, I will see what I can do here.

**Comment by nickchapman-da** [comment]
> I am curious to know how you will handle the caching of contract when a same contract need to be interpreted with different versions in the same transactions. This includes the case where the one or both of them are used as interfaces.

Let's discuss this in person, this morning!

**Comment by nickchapman-da** [comment]
Thoughts on Interaction with Caching.

A: cache an (LF) Value. Then convert to SValue on each import
B: cache SValue (like now). Then convert to a new SValue on each import.

One trickiness with B is: To make the initial conversion to an SValue (to get something to cache) requires that we have a visible type definition in hand, but in general this wont be available until we import.

**Comment by remyhaemmerle-da** [comment]
> Thoughts on Interaction with Caching.

We discuss offline and agree taking care of the caching first.

**Comment by nickchapman-da** [comment]
Review comments from Remy addressed.

Discussed with Remy how to handle the interaction of contract caching and upgrades.
We agreed to proceed with approach A: cache an (LF) Value. Then convert to SValue on each import.

I will work on this next.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: done

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 2 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala]: ok

**Review by remyhaemmerle-da — APPROVED** [review]

**Pull request merged**